internal/scheduled/runner.go

```diff
@@ -0,0 +1,170 @@
+package scheduled
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log/slog"
+	"time"
+
+	"github.com/ministryofjustice/opg-modernising-lpa/internal/donor/donordata"
+	"github.com/ministryofjustice/opg-modernising-lpa/internal/dynamo"
+	"github.com/ministryofjustice/opg-modernising-lpa/internal/identity"
+	"github.com/ministryofjustice/opg-modernising-lpa/internal/notify"
+	"github.com/ministryofjustice/opg-modernising-lpa/internal/task"
+)
+
+// errStepIgnored is returned by steps when they don't require processing
+var errStepIgnored = errors.New("step ignored")
+
+type ActionFunc func(ctx context.Context, row *Event) error
+
+type ScheduledStore interface {
+	Pop(ctx context.Context, at time.Time) (*Event, error)
+}
+
+type DonorStore interface {
+	One(ctx context.Context, pk dynamo.LpaKeyType, sk dynamo.SK) (*donordata.Provided, error)
+	Put(ctx context.Context, provided *donordata.Provided) error
+}
+
+type NotifyClient interface {
+	SendActorEmail(ctx context.Context, to, lpaUID string, email notify.Email) error
+}
+
+type Logger interface {
+	InfoContext(ctx context.Context, msg string, args ...any)
+	ErrorContext(ctx context.Context, msg string, args ...any)
+}
+
+type Waiter interface {
+	Reset()
+	Wait() error
+}
+
+type Runner struct {
+	logger       Logger
+	store        ScheduledStore
+	now          func() time.Time
+	period       time.Duration
+	donorStore   DonorStore
+	notifyClient NotifyClient
+	actions      map[Action]ActionFunc
+	waiter       Waiter
+}
+
+func NewRunner(logger Logger, store ScheduledStore, donorStore DonorStore, notifyClient NotifyClient, period time.Duration) *Runner {
+	r := &Runner{
+		logger:       logger,
+		store:        store,
+		now:          time.Now,
+		period:       period,
+		donorStore:   donorStore,
+		notifyClient: notifyClient,
+		waiter:       &waiter{backoff: time.Second, sleep: time.Sleep, maxRetries: 10},
+	}
+
+	r.actions = map[Action]ActionFunc{
+		ActionExpireDonorIdentity: r.stepCancelDonorIdentity,
+	}
+
+	return r
+}
+
+// Run the Runner, it is expected to be called in a Go routine.
+func (r *Runner) Run(ctx context.Context) error {
+	ticker := time.Tick(r.period)
+
+	for {
+		innerCtx, cancel := context.WithTimeout(ctx, r.period)
+		defer cancel()
+
+		r.logger.InfoContext(ctx, "runner step started")
+		if err := r.step(innerCtx); err != nil {
+			r.logger.ErrorContext(ctx, "runner step error", slog.Any("err", err))
+		}
+		r.logger.InfoContext(ctx, "runner step finished")
+
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-ticker:
+			continue
+		}
+	}
+}
+
+func (r *Runner) step(ctx context.Context) error {
+	r.waiter.Reset()
+
+	for {
+		row, err := r.store.Pop(ctx, r.now())
```

Comment:
to test locally you can either change the PK of the item in dynamo, or do `r.now().AddDate(0,6,0)` here